lib/internal/errors.js

```diff
@@ -0,0 +1,258 @@
+'use strict';
+
+// The whole point behind this internal module is to allow Node.js to no
+// longer be forced to treat every error message change as a semver-major
+// change. The NodeError classes here all expose a `code` property whose
+// value identifies the error.
+
+const kCode = Symbol('code');
+
+class NodeError extends Error {
+  constructor(key, ...args) {
+    super(message(key, args));
+    this[kCode] = key;
+    Error.captureStackTrace(this, NodeError);
+  }
+
+  get name() {
+    return `Error[${this[kCode]}]`;
```

Comment:
I consider the fact that we just happen to call this class `NodeError` an implementation detail. It's just an `Error` with an additional `code`. I'd rather not expose the `NodeError` part as part of the public API